### PR TITLE
Support transactions for jballerina

### PR DESCRIPTION
**Pull request by warunalakshitha**
## Purpose
Current bvm based transaction logic has been re written completely with ballerina and migrate code to run on jballerina.

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Required Balo version update
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 1]
```diff
@@ -577,16 +586,137 @@ public void visit(BLangFunction funcNode) {
...
+                                                                                            a.annotationName.value))
+                                         .collect(Collectors.toList());
+        if (participantAnnotation.isEmpty()) {
+            // function not annotated for transaction participation.
+            funcNode.body = rewrite(funcNode.body, fucEnv);
+            result = funcNode;
+            return;
+        }
+
+        // If function has transaction participant annotations it will be desugar either to beginLocalParticipant or
```
Comment (rdhananjaya) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]: Shall we do a extract method here for the participant desugaring stuff.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 2]
```diff
@@ -577,16 +586,137 @@ public void visit(BLangFunction funcNode) {
...
+        // beginRemoteParticipant function in transaction package.
+        //
+        //function beginLocalParticipant(string transactionBlockId, function () committedFunc,
+        //                               function () abortedFunc, function () trxFunc) returns error|any {
+        
+        BLangAnnotationAttachment annotation = participantAnnotation.get(0);
+        BLangBlockStmt onCommitBody = null;
+        BLangBlockStmt onAbortBody =  null;
+        funcNode.requiredParams.forEach(bLangSimpleVariable -> bLangSimpleVariable.symbol.closure = true);
+        BType trxReturnType = new BUnionType(null, new LinkedHashSet<>(Lists.of(symTable.errorType, symTable.nilType,
```
Comment (rdhananjaya) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]: We usually use BUnionType.create()

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 3]
```diff
@@ -2126,14 +2258,126 @@ boolean isParentLocked(BLangLock lock, BLangVariableReference expr) {
...
+
+        DiagnosticPos pos = transactionNode.pos;
+        BType trxReturnType = symTable.intType;
+        BType otherReturnType = symTable.nilType;
+        BLangType trxReturnNode = ASTBuilderUtil.createTypeNode(trxReturnType);
+        BLangType otherReturnNode = ASTBuilderUtil.createTypeNode(otherReturnType);
+        DiagnosticPos invPos = transactionNode.pos;
+        DiagnosticPos returnStmtPos = new DiagnosticPos(invPos.src,
+                                                        invPos.eLine, invPos.eLine, invPos.sCol, invPos.sCol);
+        BLangStatement statement = transactionNode.transactionBody.stmts
```
Comment (rdhananjaya) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]: Shall we add a comment explaining what's done here.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 4]
```diff
@@ -2126,14 +2258,126 @@ boolean isParentLocked(BLangLock lock, BLangVariableReference expr) {
...
+                .get(transactionNode.transactionBody.stmts.size() - 1);
+        if (!(statement instanceof BLangAbort) && !(statement instanceof BLangRetry)) {
```
Comment (rdhananjaya) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]: NodeKind.ABORT

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java [hunk 5]
```diff
@@ -2036,6 +2036,14 @@ public void visit(BLangTransaction transactionNode) {
             typeChecker.checkExpr(transactionNode.retryCount, env, symTable.intType);
             checkRetryStmtValidity(transactionNode.retryCount);
         }
+        env.scope.entries
```
Comment (rdhananjaya) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java]: Shall we add a comment explaining that this block going to be  desugared into a function and so on.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/types/BUnionType.java [hunk 6]
```diff
@@ -44,7 +44,7 @@
 
     private LinkedHashSet<BType> memberTypes;
 
-    private BUnionType(BTypeSymbol tsymbol, LinkedHashSet<BType> memberTypes, boolean nullable) {
+    public BUnionType(BTypeSymbol tsymbol, LinkedHashSet<BType> memberTypes, boolean nullable) {
```
Comment (rdhananjaya) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/types/BUnionType.java]: BUnionType.create method

File: stdlib/transactions/src/main/ballerina/transactions/transaction_block.bal [hunk 7]
```diff
@@ -14,6 +14,182 @@
...
+        error err = error("{ballerina}TransactionError", message =  "dynamically nested transactions are not allowed");
+        panic err;
+    }
+    TransactionContext|error txnContext;
+    string transactionId = "";
+    int|error trxResult = -1;
+    error? abortResult = ();
+    error? committedResult = ();
+    error? retryResult = ();
+    error
```
Comment (rdhananjaya) [inline stdlib/transactions/src/main/ballerina/transactions/transaction_block.bal]: merge with below line?

File: stdlib/transactions/src/main/ballerina/transactions/transaction_block.bal [hunk 8]
```diff
@@ -14,6 +14,182 @@
...
+            log:printDebug(committedResult.reason());
+        }
+    } else {
+        abortResult = trap handleAbortTransaction(transactionId, transactionBlockId, abortedFunc);
+        if (abortResult is error) {
+            log:printDebug(abortResult.reason());
+        }
+    }
+    cleanupTransactionContext(transactionBlockId);
+    if (trxResult is error) {
```
Comment (rdhananjaya) [inline stdlib/transactions/src/main/ballerina/transactions/transaction_block.bal]: This is re-throwing panic error captured in transaction block right? Shall we add a comment documenting that.

**Comment by warunalakshitha** [comment]
@rdhananjaya Thanks for the review comments. Address and fixed with latest commits.

**Pull request merged**